### PR TITLE
fix process_threads test

### PR DESCRIPTION
**Pull request by elvanderb**
Make sure that threads are started before counting them

### CODE REVIEW

File: tests/core/process.c [hunk 1]
```diff
@@ -117,9 +125,31 @@ static gboolean malloc_range_check_cb (
...
+
+  g_cond_init (&sync_data.thread_is_started);
+  g_mutex_init (&sync_data.thread_is_started_mutex);
+  g_mutex_lock (&sync_data.thread_is_started_mutex);
+  sync_data.done = done;
+
+  thread = g_thread_new ("process-test-sleeping-dummy", sleeping_dummy,
+      &sync_data);
+
+  g_cond_wait (&sync_data.thread_is_started,
```
Comment (oleavr) [inline tests/core/process.c]: Spurious wake-ups are possible, so one should always wait on a variable to change. (Look at how `g_cond_wait()` is used elsewhere.)

File: tests/core/process.c [hunk 2]
```diff
@@ -79,6 +79,13 @@ struct _TestRangeContext
   gboolean found_exact;
 };
 
+struct _TestThreadSyncData
+{
+  GMutex thread_is_started_mutex;
+  GCond thread_is_started;
```
Comment (oleavr) [inline tests/core/process.c]: Let's rename these to `mutex` and `cond`, and introduce `gboolean started;` or something like that.

File: tests/core/process.c [hunk 3]
```diff
@@ -79,6 +79,13 @@ struct _TestRangeContext
   gboolean found_exact;
 };
 
+struct _TestThreadSyncData
```
Comment (oleavr) [inline tests/core/process.c]: Missing typedef above -- should be right after the `TestRangeContext` typedef.

File: tests/core/process.c [hunk 4]
```diff
@@ -117,9 +125,31 @@ static gboolean malloc_range_check_cb (
     const GumMallocRangeDetails * details, gpointer user_data);
 #endif
 
+GThread * new_sleeping_dummy_thread_sync (volatile gboolean * done)
```
Comment (oleavr) [inline tests/core/process.c]: Return type should be on a separate line unless it's a prototype.

File: tests/core/process.c [hunk 5]
```diff
@@ -713,11 +741,14 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
 {
-  volatile gboolean * done = (gboolean *) data;
+  struct _TestThreadSyncData * sync_data = (struct _TestThreadSyncData *)data;
+  volatile gboolean * done = sync_data->done;
 
-  while (!(*done))
+  g_mutex_lock (&sync_data->thread_is_started_mutex);
+  g_cond_signal (&sync_data->thread_is_started);
+  g_mutex_unlock (&sync_data->thread_is_started_mutex);
+  while (!(* done))
```
Comment (oleavr) [inline tests/core/process.c]: There should not be a space between `*` and `done`. Also a blank line before `while` would visually separate the two things that this function does.

File: tests/core/process.c [hunk 6]
```diff
@@ -117,9 +125,31 @@ static gboolean malloc_range_check_cb (
...
+  thread = g_thread_new ("process-test-sleeping-dummy", sleeping_dummy,
+      &sync_data);
+
+  g_cond_wait (&sync_data.thread_is_started,
+      &sync_data.thread_is_started_mutex);
+  g_mutex_unlock (&sync_data.thread_is_started_mutex);
+
+  g_cond_clear (&sync_data.thread_is_started);
+  g_mutex_clear (&sync_data.thread_is_started_mutex);
+  return thread;
```
Comment (oleavr) [inline tests/core/process.c]: Blank line before `return` would look good.

File: tests/core/process.c [hunk 7]
```diff
@@ -713,11 +741,14 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
 static gpointer
 sleeping_dummy (gpointer data)
 {
-  volatile gboolean * done = (gboolean *) data;
+  struct _TestThreadSyncData * sync_data = (struct _TestThreadSyncData *)data;
```
Comment (oleavr) [inline tests/core/process.c]: - Always use typedefs.
- Missing space after cast. Cast should be dropped, though, as this is C (C++ would be a different story).

File: tests/core/process.c [hunk 8]
```diff
@@ -117,9 +125,31 @@ static gboolean malloc_range_check_cb (
     const GumMallocRangeDetails * details, gpointer user_data);
 #endif
 
+GThread * new_sleeping_dummy_thread_sync (volatile gboolean * done)
+{
+  struct _TestThreadSyncData sync_data;
```
Comment (oleavr) [inline tests/core/process.c]: Should use typedef.

File: tests/core/process.c (hunk 4 above)
Comment (oleavr) [reply tests/core/process.c]: Also, this helper function is lower level than the functions useing it, so it should be below them.

File: tests/core/process.c (hunk 7 above)
Comment (elvanderb) [reply tests/core/process.c]: I left the cast as you use the exact same construction in this file (and I prefer to use casts :D)

**Comment by elvanderb** [comment]
fixed :)

File: tests/core/process.c [hunk 9]
```diff
@@ -710,11 +718,43 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
+  g_cond_init (&sync_data.cond);
+  g_mutex_init (&sync_data.mutex);
+  g_mutex_lock (&sync_data.mutex);
+  sync_data.started = FALSE;
+  sync_data.done = done;
+
+  thread = g_thread_new ("process-test-sleeping-dummy", sleeping_dummy,
+      &sync_data);
+  
+  while (! sync_data.started)
```
Comment (oleavr) [inline tests/core/process.c]: Please remove extra space before `!`.

File: tests/core/process.c [hunk 10]
```diff
@@ -710,11 +718,43 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
 #endif
 
+static GThread *
+new_sleeping_dummy_thread_sync (volatile gboolean * done)
+{
+  TestThreadSyncData sync_data;
+  GThread * thread;
+
+  g_cond_init (&sync_data.cond);
+  g_mutex_init (&sync_data.mutex);
```
Comment (oleavr) [inline tests/core/process.c]: Fields should be initialized in the same order that they’re declared in, i.e. `mutex` and then `cond`

File: tests/core/process.c [hunk 11]
```diff
@@ -710,11 +718,43 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
+  sync_data.done = done;
+
+  thread = g_thread_new ("process-test-sleeping-dummy", sleeping_dummy,
+      &sync_data);
+  
+  while (! sync_data.started)
+    g_cond_wait (&sync_data.cond, &sync_data.mutex);
+
+  g_mutex_unlock (&sync_data.mutex);
+
```
Comment (oleavr) [inline tests/core/process.c]: Nit: I’d either drop this blank line, or add one after mutex+cond init calls, to improve visual symmetry.

File: tests/core/process.c [hunk 12]
```diff
@@ -710,11 +718,43 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
+  g_mutex_clear (&sync_data.mutex);
+
+  return thread;
+}
+
 static gpointer
 sleeping_dummy (gpointer data)
 {
-  volatile gboolean * done = (gboolean *) data;
+  TestThreadSyncData * sync_data = (TestThreadSyncData *)data;
```
Comment (oleavr) [inline tests/core/process.c]: Still missing space after cast. But let’s drop the cast and not copy the mistakes of my younger self. ;)

**Comment by elvanderb** [comment]
done

File: tests/core/process.c [hunk 13]
```diff
@@ -710,11 +718,42 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
 {
-  volatile gboolean * done = (gboolean *) data;
+  TestThreadSyncData * sync_data = data;
+  volatile gboolean * done = sync_data->done;
 
+  g_mutex_lock (&sync_data->mutex);
+  sync_data->started = TRUE;
+  g_cond_signal (&sync_data->cond);
+  g_mutex_unlock (&sync_data->mutex);
+
```
Comment (oleavr) [inline tests/core/process.c]: Trailing whitespace.

File: tests/core/process.c [hunk 14]
```diff
@@ -710,11 +718,42 @@ store_export_address_if_mach_msg (const GumExportDetails * details,
...
+  g_mutex_init (&sync_data.mutex);
+  g_mutex_lock (&sync_data.mutex);
+  g_cond_init (&sync_data.cond);
+  sync_data.started = FALSE;
+  sync_data.done = done;
+
+  thread = g_thread_new ("process-test-sleeping-dummy", sleeping_dummy,
+      &sync_data);
+
```
Comment (oleavr) [inline tests/core/process.c]: Trailing whitespace.

**Comment by oleavr** [comment]
Let me merge this branch and make the final adjustments.

**Comment by oleavr** [comment]
Landed in 4336ac9. Thanks!

**Pull request closed without merging**